### PR TITLE
feat: Add codespaces/gitpod support

### PR DESCRIPTION
**Pull request by michaelbrewer**
## Description of your changes

- Add configuration for codespaces and gitpod
- Set node version via `.nvmrc` to latest v14
- Set npm version to `npm@next-8`

### How to verify this change

- Create new codespaces workspace from PR
- Open PR in [GitPod](https://gitpod.io/#https://github.com/gyft/aws-lambda-powertools-typescript/tree/feat-codespaces)
- And run `npm test` or pushing a commit to check if pre-commits work.

### Related issues, RFCs

- #115
- #426

### PR status

***Is this ready for review?:*** NO  
***Is it a breaking change?:*** NO

## Checklist

- [X] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
- [X] I have performed a *self-review* of my own code
- [X] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [X] I have made corresponding changes to the *documentation*
- [X] My changes generate *no new warnings*
- [X] The *code coverage* hasn't decreased
- [X] I have *added tests* that prove my change is effective and works
- [X] New and existing *unit tests pass* locally and in Github Actions
- [X] Any *dependent changes have been merged and published* in downstream module
- [X] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: .devcontainer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# See here for image contents: https://github.com/microsoft/vscode-dev-containers/blob/v0.212.0/containers/javascript-node/.devcontainer/base.Dockerfile
+# [Choice] Node.js version (use -bullseye variants on local arm64/Apple Silicon): 16, 14, 12, 16-bullseye, 14-bullseye, 12-bullseye, 16-buster, 14-buster, 12-buster
+ARG VARIANT="16-bullseye"
```
Comment (flochaz) [inline .devcontainer/Dockerfile:3]: We should stick to the latest node version supported by AWS Lambda (14 for now unfortunately). We need to have clearer guidelines on this ... Discussion in progress https://github.com/awslabs/aws-lambda-powertools-typescript/discussions/331

Comment (michaelbrewer) [reply .devcontainer/Dockerfile:3]: sure. i can keep it at node 14 level

**Comment by flochaz** [comment]
LGTM but will wait for a consensus to be reached here  #331 before merging.

**Comment by saragerion** [comment]
Thanks for opening this PR @michaelbrewer! I really like that we will be supporting both codespaces and gitpod. CC; @loujaybee <3 
As Florian mentioned, we need to fine-tune some decisions around versioning first, but looking forward to merging this one.

**Comment by michaelbrewer** [comment]
> Thanks for opening this PR @michaelbrewer! I really like that we will be supporting both codespaces and gitpod. CC; @loujaybee <3 As Florian mentioned, we need to fine-tune some decisions around versioning first, but looking forward to merging this one.

Sounds good. I can see how to spin out the version pinning to one place. I see that the e2e is focusing on node v14, which makes 100% sense. 

It is really awesome using Gitpod (for free) on lightweight edits ;)

**Comment by dreamorosi** [comment]
Discussion #331 has been resolved and PR #392 that expands the contributing section has been merged.

Ideally this PR should take in account the new node/npm versioning in this repo and uncomment [this line](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/CONTRIBUTING.md?plain=1#L111) in the `CONTRIBUTING` doc before being merged.

**Comment by michaelbrewer** [comment]
@flochaz @dreamorosi  - ok it is ready to look at again. there is an issue with the `examples/cdk` but i will do that on a seperate PR

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+RUN apt-get update && export DEBIAN_FRONTEND=noninteractive \
+    && apt-get -y install --no-install-recommends bash-completion
+
+# [Optional] Uncomment if you want to install an additional version of node using nvm
+# ARG EXTRA_NODE_VERSION=10
+# RUN su node -c "umask 0002 && ./usr/local/share/nvm/nvm.sh && nvm install ${EXTRA_NODE_VERSION}"
+
+# [Optional] Uncomment if you want to install more global node modules
+# RUN su node -c "npm install -g <your-package-list-here>"
+RUN su node -c "npm install -g npm-check-updates"
```
Comment (dreamorosi) [inline .devcontainer/Dockerfile:18]: We should install `npm@next-8` here like the `CONTRIBUTING.md` doc says.

File: .gitpod.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+tasks:
+  - init: nvm install; nvm use; npm ci; npm run lerna-ci; npm run init-environment
```
Comment (dreamorosi) [inline .gitpod.yml:2]: Same here, `node@14.x` comes with `npm@6.x` if I'm not mistaken but we're targeting `npm@8.x` which follows the `@next-8` on the npm registry.

**Review by dreamorosi — CHANGES_REQUESTED** [review]

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (michaelbrewer) [reply .devcontainer/Dockerfile:18]: Oh ok. Would it be strange to have npm@8 when npm@6 is installed by default for all node@14 users ?

File: .gitpod.yml (hunk 3 above)
Comment (michaelbrewer) [reply .gitpod.yml:2]: Yes node@14 does include npm@6 so I would expect that to be the norm until people move to node@16?

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (michaelbrewer) [reply .devcontainer/Dockerfile]: ok done @dreamorosi

File: .gitpod.yml (hunk 3 above)
Comment (michaelbrewer) [reply .gitpod.yml]: Ok done.

Maybe we can have a helper make file like this?

```Makefile
dev:
	  nvm install
	  nvm use
	  npm i -g npm@next-8
	  npm ci
	  npm run lerna-ci
	  npm run init-environment
```

**Comment by flochaz** [comment]
Lint and build checks failing due to https://github.com/awslabs/aws-lambda-powertools-typescript/issues/494 which should be fixed by https://github.com/awslabs/aws-lambda-powertools-typescript/pull/498

**Comment by michaelbrewer** [comment]
@flochaz - ok i have merged in #498 changes

**Review by dreamorosi — APPROVED** [review]

**Review by flochaz — APPROVED** [review]

**Pull request merged**